<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

# Authentication Routes

Stand alone functions for users who prefer a functional programming approach

## Full Auth Route - username and password authentication

In [1]:
#| echo: false
#| output: asis
show_doc(get_full_auth)

---

[source](https://github.com/jaewilson07/domo_library/blob/main/domolibrary/DomoAuth.py#L17){target="_blank" style="float:right; font-size:smaller"}

### get_full_auth

>      get_full_auth (domo_instance:str, domo_username:str, domo_password:str,
>                     session:Optional[aiohttp.client.ClientSession]=None)

uses username and password authentication to retrieve a full_auth access token

#### Sample Implementations of get_full_auth

In [2]:
import os

res = await get_full_auth(
    domo_instance="domo-dojo",
    domo_username="test@test.com",
    domo_password="fake password",
)
res

ResponseGetData(status=200, response={'success': False, 'reason': 'INVALID_CREDENTIALS'}, is_success=True, auth=None)

In [ ]:
res = await get_full_auth(domo_instance="test",
                          domo_username="fake@test.com", 
                          domo_password="fake password")
res

ResponseGetData(status=403, response='Forbidden', is_success=False, auth=None)

The 403 response is expected because test.domo.com presumeably does not exist or access if forbidden.

## Developer Auth Route - client_id and secret authentication

In [3]:
#| echo: false
#| output: asis
show_doc(get_developer_auth)

---

[source](https://github.com/jaewilson07/domo_library/blob/main/domolibrary/DomoAuth.py#L49){target="_blank" style="float:right; font-size:smaller"}

### get_developer_auth

>      get_developer_auth (domo_client_id:str, domo_client_secret:str,
>                          session:Optional[aiohttp.client.ClientSession]=None)

only use for authenticating against apis documented under developer.domo.com

#### Sample Implementations of get_developer_auth

The 401 response is expected because we are using invalid credentials

In [ ]:
await get_developer_auth(domo_client_id="test_id",
                         domo_client_secret="test_secret"
                         )

ResponseGetData(status=401, response='Unauthorized', is_success=False, auth=None)

## Access Token Auth Route - access_token authentication

This access_token based authentication assumes the user has been provided a valid access token from Domo > Admin > Authentication > Access Token so authentication routes are actually not required.

Per the Domo JavaCLI implementation, users can test the validity of the access_token agains the 'me' API

!! Note about the Me API !!
It appears that access_token authentication will direct the

In [4]:
#| echo: false
#| output: asis
show_doc(test_access_token)

---

[source](https://github.com/jaewilson07/domo_library/blob/main/domolibrary/DomoAuth.py#L76){target="_blank" style="float:right; font-size:smaller"}

### test_access_token

>      test_access_token (domo_access_token:str, domo_instance:str,
>                         session:Optional[aiohttp.client.ClientSession]=None)

will attempt to validate against the 'me' API.
This is the same authentication test the Domo Java CLI uses.

#### Sample implementation of [`test_access_token`](https://jaewilson07.github.io/domo_library/domoauth.html#test_access_token)

In [5]:
import os

await test_access_token(
    domo_instance="domo-dojo",
    domo_access_token=os.environ['DOMO_DOJO_ACCESS_TOKEN']
)

ResponseGetData(status=200, response={'id': 1893952720, 'invitorUserId': 587894148, 'displayName': 'Jae Wilson', 'userName': 'jae@onyxreporting.com', 'emailAddress': 'jae@onyxreporting.com', 'avatarKey': 'c605f478-0cd2-4451-9fd4-d82090b71e66', 'accepted': True, 'userType': 'USER', 'modified': 1651692505000, 'created': 1588960518, 'role': 'Admin', 'rights': 63, 'active': True, 'pending': False, 'systemUser': False, 'anonymous': False}, is_success=True, auth=None)

# Classes

Different Domo Auth classes will have a variety of required vs optional parameters. To avoid multiple initialization and post_intialization statements, we mix multiple classes together such that classes with optional parameters are mixed in before classes with required parameters.

In [6]:
#| echo: false
#| output: asis
show_doc(DomoAuth)

---

[source](https://github.com/jaewilson07/domo_library/blob/main/domolibrary/DomoAuth.py#L141){target="_blank" style="float:right; font-size:smaller"}

### DomoAuth

>      DomoAuth (domo_instance:str, token:Optional[str]=None,
>                token_name:Optional[str]=None, user_id:Optional[str]=None,
>                auth_header:dict=<factory>,
>                url_manual_login:Optional[str]=None)

abstract DomoAuth class

In [ ]:
# Attributes of `DomoAuth`
domo_instance = "test"
[attr for attr in dir(DomoAuth(domo_instance)) if not attr.startswith("__")]

['auth_header',
 'domo_instance',
 'generate_auth_header',
 'get_auth_token',
 'set_manual_login',
 'token',
 'token_name',
 'url_manual_login',
 'user_id']

In [ ]:
# validate can print manual login link
domo_instance = "test"

da = DomoAuth(domo_instance)
da.url_manual_login

'https://test.domo.com/auth/index?domoManualLogin=true'

## DomoAuth Error Classes

In [7]:
#| echo: false
#| output: asis
show_doc(InvalidInstanceError)

---

[source](https://github.com/jaewilson07/domo_library/blob/main/domolibrary/DomoAuth.py#L174){target="_blank" style="float:right; font-size:smaller"}

### InvalidInstanceError

>      InvalidInstanceError (status:Optional[int]=None, message='invalid
>                            instance', domo_instance:Optional[str]=None)

return if invalid domo_instance sent to API

In [8]:
#| echo: false
#| output: asis
show_doc(InvalidCredentialsError)

---

[source](https://github.com/jaewilson07/domo_library/blob/main/domolibrary/DomoAuth.py#L161){target="_blank" style="float:right; font-size:smaller"}

### InvalidCredentialsError

>      InvalidCredentialsError (status:Optional[int]=None, message='invalid
>                               credentials', domo_instance:Optional[str]=None)

return invalid credentials sent to API

|    | **Type** | **Default** | **Details** |
| -- | -------- | ----------- | ----------- |
| status | Optional | None | API request status |
| message | str | invalid credentials |  |
| domo_instance | Optional | None |  |

## DomoFullAuth

In [9]:
#| echo: false
#| output: asis
show_doc(DomoFullAuth)

---

[source](https://github.com/jaewilson07/domo_library/blob/main/domolibrary/DomoAuth.py#L196){target="_blank" style="float:right; font-size:smaller"}

### DomoFullAuth

>      DomoFullAuth (domo_instance:str, domo_username:str, domo_password:str,
>                    token:Optional[str]=None, token_name:Optional[str]=None,
>                    user_id:Optional[str]=None, auth_header:dict=<factory>,
>                    url_manual_login:Optional[str]=None)

use for full authentication token

#### sample implementations of DomoFullAuth

In [ ]:
try:
    full_auth = DomoFullAuth(domo_instance="domo-dojo",
                             domo_username="test12@domo.com",
                             domo_password="test1234")
    res = await full_auth.get_auth_token()

except InvalidCredentialsError as e:
    print(e)

Status 200 - INVALID_CREDENTIALS at domo-dojo


In [ ]:
try:
    full_auth = DomoFullAuth(domo_instance="test",
                             domo_username="test12@domo.com",
                             domo_password="test1234")
    await full_auth.get_auth_token()
except InvalidInstanceError as e:
    print(e)

Status 403 - INVALID INSTANCE at test


## DomoTokenAuth

In [10]:
#| echo: false
#| output: asis
show_doc(DomoTokenAuth)

---

[source](https://github.com/jaewilson07/domo_library/blob/main/domolibrary/DomoAuth.py#L250){target="_blank" style="float:right; font-size:smaller"}

### DomoTokenAuth

>      DomoTokenAuth (domo_instance:str, domo_access_token:str,
>                     token:Optional[str]=None, token_name:Optional[str]=None,
>                     user_id:Optional[str]=None, auth_header:dict=<factory>,
>                     url_manual_login:Optional[str]=None)

use for access_token authentication.
Tokens are generated in domo > admin > access token
Necessary in cases where direct sign on is not permitted

#### Sample implementation of DomoTokenAuth

In [ ]:
try:
    domo_auth = DomoTokenAuth(domo_instance="domo-dojo",
                              domo_access_token='fake password')
    token = await domo_auth.get_auth_token()
    print(domo_auth)

except InvalidCredentialsError as e:
    print(e)

Status 401 - Unauthorized at domo-dojo


## DomoDeveloperAuth

In [11]:
#| echo: false
#| output: asis
show_doc(DomoDeveloperAuth)

---

[source](https://github.com/jaewilson07/domo_library/blob/main/domolibrary/DomoAuth.py#L302){target="_blank" style="float:right; font-size:smaller"}

### DomoDeveloperAuth

>      DomoDeveloperAuth (domo_client_id:str, domo_client_secret:str)

use for full authentication token

#### Sample implementations of DomoDeveloperAuth

In [ ]:
domo_client_id = "test_client"
domo_client_secret = "test_secret"

try:
    domo_auth = DomoDeveloperAuth(domo_client_id, domo_client_secret)
    await domo_auth.get_auth_token()
except InvalidCredentialsError as e:
    print(e)

Status 401 - Unauthorized
